### PR TITLE
Add guard for multiple node unmarshal

### PR DESCRIPTION
**Pull request by liusanchuan**
Fixes unmarshalling a node with multiple left and right children

## Description

Add a check for unmarshalling a node when the input byte array with multiple left or right childs.
[Mentioned in Here](https://github.com/NethermindEth/juno/blob/develop/core/trie/node.go#:~:text=//%20TODO%3A%20Implement%20and,will%20still%20succeed.)

## Changes:

- Add left and right flag to protect there only at most 1 left or right child when marshalling.
- Add unit test for this part.

## Types of changes

_Leave only items that describe your changes, remove the rest. Remove this line too._

- Bugfix (non-breaking change which fixes an issue)

## Testing

**Requires testing**: Yes

**Did you write tests??**: Yes

## Documentation

**If this requires a documentation update, did you add one?** No

### CODE REVIEW

File: core/trie/node_test.go [hunk 1]
```diff
@@ -88,6 +88,64 @@ func TestNodeMarshalAndUnmarshalBinary(t *testing.T) {
...
+				marshalBin: flexiableMarshal(value, []bitset.BitSet{*path1, *path2}, []bitset.BitSet{*path3, *path4}),
+				errStr:     "malformed node: multiple left childs are not support",
+			},
+		}
+		for _, test := range tests {
+			t.Run(test.name, func(t *testing.T) {
+				unmarshalled := new(Node)
+				err = unmarshalled.UnmarshalBinary(test.marshalBin)
+				if err.Error() != test.errStr {
+					t.Fatalf("expected error not right")
```
Comment (joshklop) [inline core/trie/node_test.go:144]: Nitpick. We don't need to fail the entire test suite if this test fails, so let's use `Errorf` instead of `Fatalf`. Let's also make the error message slightly more informative.
```suggestion
					t.Errorf("expected error not right: got %s, wanted ErrMalformedNode", err)
```

File: core/trie/node.go [hunk 2]
```diff
@@ -109,9 +110,17 @@ func (n *Node) UnmarshalBinary(data []byte) error {
...
+				pathP = &(n.left)
+			} else {
+				return ErrMalformedNode{"multiple left childs are not support"}
+			}
 		case 'r':
-			pathP = &(n.right)
+			if n.right == nil {
+				pathP = &(n.right)
+			} else {
+				return ErrMalformedNode{"multiple right childs are not support"}
```
Comment (joshklop) [inline core/trie/node.go:122]:
```suggestion
				return ErrMalformedNode{"multiple right children are not supported"}
```

File: core/trie/node.go [hunk 3]
```diff
@@ -109,9 +110,17 @@ func (n *Node) UnmarshalBinary(data []byte) error {
 		var pathP **bitset.BitSet
 		switch head {
 		case 'l':
-			pathP = &(n.left)
+			if n.left == nil {
+				pathP = &(n.left)
+			} else {
+				return ErrMalformedNode{"multiple left childs are not support"}
```
Comment (joshklop) [inline core/trie/node.go:116]:
```suggestion
				return ErrMalformedNode{"multiple left children are not supported"}
```

**Review by joshklop — CHANGES_REQUESTED** [review]
Amazing work! Your contribution is greatly appreciated.

File: core/trie/node_test.go [hunk 4]
```diff
@@ -88,6 +88,64 @@ func TestNodeMarshalAndUnmarshalBinary(t *testing.T) {
...
+			},
+			{
+				name:       "node with 2 left and 2 right child",
+				marshalBin: flexiableMarshal(value, []bitset.BitSet{*path1, *path2}, []bitset.BitSet{*path3, *path4}),
+				errStr:     "malformed node: multiple left childs are not support",
+			},
+		}
+		for _, test := range tests {
+			t.Run(test.name, func(t *testing.T) {
+				unmarshalled := new(Node)
```
[the remark this answers is not in the document]
Comment (liusanchuan) [reply core/trie/node_test.go]: @joshklop  Thanks for your detail correction, I have fix these error in next commit, Please help to review :)

File: core/trie/node.go (hunk 3 above)
Comment (liusanchuan) [reply core/trie/node.go]: Sorry for make so many spell error.

File: core/trie/node_test.go [hunk 5]
```diff
@@ -88,6 +89,61 @@ func TestNodeMarshalAndUnmarshalBinary(t *testing.T) {
...
+	t.Run("Test edge case of malformed node", func(t *testing.T) {
+		value, err := new(felt.Felt).SetRandom()
+		if err != nil {
+			t.Fatalf("expected no error but got %s", err)
+		}
+		path1 := bitset.FromWithLength(44, []uint64{11})
+		path2 := bitset.FromWithLength(22, []uint64{22})
+		path3 := bitset.FromWithLength(22, []uint64{33})
+		path4 := bitset.FromWithLength(22, []uint64{44})
+		flexiableMarshal := func(val *felt.Felt, left []bitset.BitSet, right []bitset.BitSet) []byte {
```
Comment (jelilat) [inline core/trie/node_test.go:101]:
```suggestion
		flexibleMarshal := func(val *felt.Felt, left []bitset.BitSet, right []bitset.BitSet) []byte {
```

File: core/trie/node_test.go [hunk 6]
```diff
@@ -88,6 +89,61 @@ func TestNodeMarshalAndUnmarshalBinary(t *testing.T) {
...
+		}
+
+		tests := [...]struct {
+			name       string
+			marshalBin []byte
+			errStr     string
+		}{
+			{
+				name:       "node with only 2 left children",
+				marshalBin: flexiableMarshal(value, []bitset.BitSet{*path1, *path2}, []bitset.BitSet{}),
```
Comment (jelilat) [inline core/trie/node_test.go:126]:
```suggestion
				marshalBin: flexibleMarshal(value, []bitset.BitSet{*path1, *path2}, []bitset.BitSet{}),
```

File: core/trie/node_test.go [hunk 7]
```diff
@@ -88,6 +89,61 @@ func TestNodeMarshalAndUnmarshalBinary(t *testing.T) {
...
+			},
+			{
+				name:       "node with only 2 right children",
+				marshalBin: flexiableMarshal(value, []bitset.BitSet{}, []bitset.BitSet{*path3, *path4}),
```
Comment (jelilat) [inline core/trie/node_test.go:130]:
```suggestion
				marshalBin: flexibleMarshal(value, []bitset.BitSet{}, []bitset.BitSet{*path3, *path4}),
```

File: core/trie/node_test.go [hunk 8]
```diff
@@ -88,6 +89,61 @@ func TestNodeMarshalAndUnmarshalBinary(t *testing.T) {
...
+			},
+			{
+				name:       "node with 2 left and 2 right children",
+				marshalBin: flexiableMarshal(value, []bitset.BitSet{*path1, *path2}, []bitset.BitSet{*path3, *path4}),
```
Comment (jelilat) [inline core/trie/node_test.go:134]:
```suggestion
				marshalBin: flexibleMarshal(value, []bitset.BitSet{*path1, *path2}, []bitset.BitSet{*path3, *path4}),
```

**Review by jelilat — CHANGES_REQUESTED** [review]
Don't forget to remove the [TODO](https://github.com/NethermindEth/juno/pull/508/files#diff-84877d01f33f876f2ddf0ee4bd92114cc87d06cb18c0dba5bbe0ada34b537cc1R91-L95) comment here as well

**Comment by liusanchuan** [comment]
@jelilat Thanks for your point out, I have update the TODO and fixed the spell error. I think I need to install a spell check to help me.

File: core/trie/node.go (hunk 3 above)
Comment (joshklop) [reply core/trie/node.go]: No worries! This is what reviews are for.

File: core/trie/node.go [hunk 9]
```diff
@@ -88,30 +88,33 @@ func (n *Node) MarshalBinary() ([]byte, error) {
...
 		head, err := stream.ReadByte()
 		if err != nil {
 			return err
 		}
-
 		var pathP **bitset.BitSet
 		switch head {
 		case 'l':
-			pathP = &(n.left)
+			if n.left == nil {
```
Comment (omerfirmak) [inline core/trie/node.go:107]: Actually, you can move this check after the `switch` and do it only once.

something like below should work
``` golang
if *pathP != nil {
    return err
}
```

File: core/trie/node.go [hunk 10]
```diff
@@ -88,30 +88,33 @@ func (n *Node) MarshalBinary() ([]byte, error) {
...
-	//		child. However if the UnmarshalBinary is called with multiple left and/or right child
-	//		in any order then UnmarshalBinary will still succeed.
 	if len(data) < felt.Bytes {
 		return ErrMalformedNode{"size of input data is less than felt size"}
 	}
 	n.value = new(felt.Felt).SetBytes(data[:felt.Bytes])
 	data = data[felt.Bytes:]
 
 	stream := bytes.NewReader(data)
+
```
Comment (omerfirmak) [inline core/trie/node.go:98]: superfluous diff

File: core/trie/node.go (hunk 9 above)
Comment (liusanchuan) [reply core/trie/node.go:107]: Hi, @omerfirmak , your suggestion seems not work in this context. As `*pathP` is point to one of the Node's child,  after enter in case `l` or `r`, it's reference can be decided.

Comment (omerfirmak) [reply core/trie/node.go:107]: It should work. Did you try it, can I see your code?

Comment (liusanchuan) [reply core/trie/node.go:107]: Hi, @omerfirmak, I try to follow your mind, but the code still not work.
Please check [my implement ](https://github.com/liusanchuan/juno/tree/test-pathP-efficient-check).

Comment (liusanchuan) [reply core/trie/node.go]: Solved in the efficient way as you suggested, as [here](https://github.com/NethermindEth/juno/pull/508/commits/d48cbf3ea4f457fbfddc8468478e5e99dc7803b7)

**Review by omerfirmak — APPROVED** [review]
LGTM, should be squash merged tho.

**Comment by liusanchuan** [comment]
@jelilat @joshklop I have squashed all my commits, please have a re-review. Thanks

**Comment by omerfirmak** [comment]
@IronGauntlets, can you force merge this? @jelilat wont be able to review soon probably

**Review by jelilat — APPROVED** [review]

**Comment by jelilat** [comment]
@liusanchuan Please rebase the branch and feel free to merge

**Comment by liusanchuan** [comment]
It's my second PR succcessfully merged :) 
Thanks for your guys, I've leaned spell checking, rebase and squash, golang effieiency optimization. 
Juno is a amazing place, I just contribute several lines of code, but get so much tutorial for yours. For each respose, thank you again🥇 ! 
Hope I can do better and contribute more!
